src/main/kotlin/org/pkl/lsp/completion/UnqualifiedAccessCompletionProvider.kt

```diff
@@ -239,48 +256,382 @@ class UnqualifiedAccessCompletionProvider(private val project: Project) : Comple
+        collector += createPropertyAmendElement(propertyName, amendingPropertyType, property)
+      }
+      collector += createPropertyAssignElement(propertyName, propertyType, property)
+    }
+  }
+
+  private fun addMappingCompletions(
+    alreadyDefinedProperties: Set<String>,
+    thisType: Type.Class,
+    base: PklBaseModule,
+    collector: MutableList<CompletionItem>,
+    context: PklProject?,
+  ) {
+
+    val keyType = thisType.typeArguments[0]
+    val valueType = thisType.typeArguments[1]
+
+    val amendedValueType = valueType.amended(base, context)
+    if (amendedValueType != Type.Nothing && amendedValueType != Type.Unknown) {
+      val amendingValueType = valueType.amending(base, context)
+      if ("default" !in alreadyDefinedProperties) {
+        collector +=
+          createPropertyAmendElement("default", amendingValueType, base.mappingDefaultProperty)
+      }
+      collector += createEntryAmendElement(keyType, amendingValueType, base)
+    }
+    if ("default" !in alreadyDefinedProperties) {
+      collector += createPropertyAssignElement("default", valueType, base.mappingDefaultProperty)
+    }
+    collector += createEntryAssignElement(keyType, valueType, base)
+  }
+
+  private fun addListingCompletions(
+    alreadyDefinedProperties: Set<String>,
+    thisType: Type.Class,
+    base: PklBaseModule,
+    collector: MutableList<CompletionItem>,
+    context: PklProject?,
+  ) {
+
+    val elementType = thisType.typeArguments[0]
+
+    val amendedElementType = elementType.amended(base, context)
+    if (amendedElementType != Type.Nothing && amendedElementType != Type.Unknown) {
+      val amendingElementType = elementType.amending(base, context)
+      if ("default" !in alreadyDefinedProperties) {
+        collector +=
+          createPropertyAmendElement("default", amendingElementType, base.listingDefaultProperty)
+      }
+      collector += createPropertyAmendElement("new", amendingElementType, null)
+    }
+    if ("default" !in alreadyDefinedProperties) {
+      collector += createPropertyAssignElement("default", elementType, base.listingDefaultProperty)
+    }
+  }
+
+  private fun createPropertyAmendElement(
+    propertyName: String,
+    propertyType: Type,
+    propertyCtx: PklProperty?,
+  ): CompletionItem {
+    return CompletionItem("$propertyName {…}").apply {
+      insertTextFormat = InsertTextFormat.Snippet
+      insertText =
+        """
+        ${propertyCtx.modifiersStr}$propertyName {
+          ${'$'}{1:body}
+        }
+        """
+          .trimIndent()
+      detail = propertyType.render()
+      kind = CompletionItemKind.Property
+    }
+  }
+
+  private fun createPropertyAssignElement(
+    propertyName: String,
+    propertyType: Type,
+    propertyCtx: PklProperty,
+  ): CompletionItem {
+    return CompletionItem("$propertyName = ").apply {
+      insertTextFormat = InsertTextFormat.Snippet
+      insertText = "${propertyCtx.modifiersStr}$propertyName = \${1:value}"
```

Comment:
Here, and everywhere else: I think the placeholder text adds a little more noise than is helpful.

```suggestion
      insertText = "${propertyCtx.modifiersStr}$propertyName = $1"
```